### PR TITLE
add EXAONE model support

### LINKED ISSUE #8991: Feature Request: Add support for EXAONE-3.0-7.8B-Instruct model
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

please add new model [EXAONE-3.0-7.8B-Instruct](https://huggingface.co/LGAI-EXAONE/EXAONE-3.0-7.8B-Instruct)

### Motivation

llama.cpp for new model

### Possible Implementation

_No response_

### PR DESCRIPTION
**Pull request by mscheong01**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

closes #8991

This is a new model released by LG AI with good english & Korean capabilities
image

Test run:
image

### CODE REVIEW

**Comment by Nexesenex** [comment]
Hey, thanks for bringing this model up!
You might want to merge master on your PR, and fix :
src/llama.cpp:
	7800: Trailing whitespace
:)

File: src/llama.cpp [hunk 1]
```diff
@@ -7673,7 +7736,7 @@ static int llama_model_load(const std::string & fname, llama_model & model, llam
 
         llm_load_print_meta(ml, model);
 
-        if (model.vocab.type != LLAMA_VOCAB_TYPE_NONE &&
+        if (model.vocab.type != LLAMA_VOCAB_TYPE_NONE &&
```
Comment (ggerganov) [inline src/llama.cpp:7739]:
```suggestion
        if (model.vocab.type != LLAMA_VOCAB_TYPE_NONE &&
```

**Review by ggerganov — APPROVED** [review]

File: convert_hf_to_gguf.py [hunk 2]
```diff
@@ -3595,6 +3598,76 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
...
+        # attention_dropout_rate = hparams["attention_dropout"]
+        # ignore for now as EXAONE-3.0-7.8B-Instruct embed_dropout is 0.0
+        # embed_dropout_rate = hparams["embed_dropout"]
+        self.gguf_writer.add_embedding_length(embed_dim)
+        self.gguf_writer.add_head_count(num_heads)
+        self.gguf_writer.add_head_count_kv(num_kv_heads)
+        self.gguf_writer.add_context_length(max_position_embeddings)
+        self.gguf_writer.add_layer_norm_rms_eps(layer_norm_eps)
+        self.gguf_writer.add_feed_forward_length(intermediate_size)
+        self.gguf_writer.add_block_count(num_layers)
```
Comment (compilade) [inline convert_hf_to_gguf.py:3628]: `self.gguf_writer.add_file_type(self.ftype)` is missing. This is necessary to make the output of `convert_hf_to_gguf.py` exactly the same as the output of `llama-quantize` from a `f32` GGUF model of this architecture.

Comment (mscheong01) [reply convert_hf_to_gguf.py:3628]: Thanks!

File: src/llama.cpp [hunk 3]
```diff
@@ -5442,6 +5472,9 @@ static void llm_load_vocab(
             } else if (
                 tokenizer_pre == "codeshell") {
                 vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_CODESHELL;
+            } else if (
+                tokenizer_pre == "exaone") {
+                vocab.type_pre = LLAMA_VOCAB_PRE_TYPE_EXAONE;
```
Comment (compilade) [inline src/llama.cpp:5477]: This also needs to be handled in `src/llama-vocab.cpp`. I think the pre-tokenizer is similar to starcoder, and so this should be added somewhere over there:

```diff
diff --git a/src/llama-vocab.cpp b/src/llama-vocab.cpp
index 11fffce9..17deefaa 100644
--- a/src/llama-vocab.cpp
+++ b/src/llama-vocab.cpp
@@ -388,6 +388,7 @@ struct llm_tokenizer_bpe {
             case LLAMA_VOCAB_PRE_TYPE_COMMAND_R:
             case LLAMA_VOCAB_PRE_TYPE_SMOLLM:
             case LLAMA_VOCAB_PRE_TYPE_CODESHELL:
+            case LLAMA_VOCAB_PRE_TYPE_EXAONE:
                 regex_exprs = {
                     "\\p{N}",
                     "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)",
```

**Review by compilade — APPROVED** [review]
Looks good to me.

**Comment by chris-jaehoon** [comment]
thanks for your service!

but I have a problem in quantizing this model. :(

Can you update this feature?

**Comment by mscheong01** [comment]
@chris-jaehoon I previously checked that the Q4_0 quant works & runs correctly. Can you provide more information on your failed attempt?

**Comment by freddydopfel** [comment]
Has anyone uploaded the working model back to Ollama?

**Pull request merged**